ui/ducks/send/helpers.js

```diff
@@ -0,0 +1,295 @@
+import { ETH, GWEI } from '../../helpers/constants/common';
+import { calcTokenAmount } from '../../helpers/utils/token-util';
+import { MIN_GAS_LIMIT_HEX } from '../../pages/send/send.constants';
+import {
+  addGasBuffer,
+  generateERC20TransferData,
+  generateERC721TransferData,
+  getAssetTransferData,
+} from '../../pages/send/send.utils';
+import { getGasPriceInHexWei } from '../../selectors';
+import { estimateGas } from '../../store/actions';
+
+export async function estimateGasLimitForSend({
+  selectedAddress,
+  value,
+  gasPrice,
+  sendToken,
+  to,
+  data,
+  isNonStandardEthChain,
+  chainId,
+  gasLimit,
+  ...options
+}) {
+  let isSimpleSendOnNonStandardNetwork = false;
+
+  // blockGasLimit may be a falsy, but defined, value when we receive it from
+  // state, so we use logical or to fall back to MIN_GAS_LIMIT_HEX. Some
+  // network implementations check the gas parameter supplied to
+  // eth_estimateGas for validity. For this reason, we set token sends
+  // blockGasLimit default to a higher number. Note that the current gasLimit
+  // on a BLOCK is 15,000,000 and will be 30,000,000 on mainnet after London.
+  // Meanwhile, MIN_GAS_LIMIT_HEX is 0x5208.
+  let blockGasLimit = MIN_GAS_LIMIT_HEX;
+  if (options.blockGasLimit) {
+    blockGasLimit = options.blockGasLimit;
+  } else if (sendToken) {
+    blockGasLimit = GAS_LIMITS.BASE_TOKEN_ESTIMATE;
+  }
+
+  // The parameters below will be sent to our background process to estimate
+  // how much gas will be used for a transaction. That background process is
+  // located in tx-gas-utils.js in the transaction controller folder.
+  const paramsForGasEstimate = { from: selectedAddress, value, gasPrice };
+
+  if (sendToken) {
+    if (!to) {
+      // if no to address is provided, we cannot generate the token transfer
+      // hexData. hexData in a transaction largely dictates how much gas will
+      // be consumed by a transaction. We must use our best guess, which is
+      // represented in the gas shared constants.
+      return GAS_LIMITS.BASE_TOKEN_ESTIMATE;
+    }
+    paramsForGasEstimate.value = '0x0';
+
+    // We have to generate the erc20/erc721 contract call to transfer tokens in
+    // order to get a proper estimate for gasLimit.
+    paramsForGasEstimate.data = getAssetTransferData({
+      sendToken,
+      fromAddress: selectedAddress,
+      toAddress: to,
+      amount: value,
+    });
+
+    paramsForGasEstimate.to = sendToken.address;
+  } else {
+    if (!data) {
+      // eth.getCode will return the compiled smart contract code at the
+      // address. If this returns 0x, 0x0 or a nullish value then the address
+      // is an externally owned account (NOT a contract account). For these
+      // types of transactions the gasLimit will always be 21,000 or 0x5208
+      const { isContractAddress } = to
+        ? await readAddressAsContract(global.eth, to)
+        : {};
+      if (!isContractAddress && !isNonStandardEthChain) {
```

Comment:
Is something like this better?

```suggestion
      if (!isContractAddress) {
        if(isNonStandardEthChain) {
          isSimpleSendOnNonStandardNetwork = true;
        }
        else {
          return GAS_LIMITS.SIMPLE;
        }
      }
```

Reply:
I like what you have, but i'm hesitant to change it as this function simply moved unchanged from its previous location.